packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -382,17 +358,77 @@ export function makeVirtualObjectManager(
           incRefCount(vref);
         } else {
           // exported non-virtual object: Remotable
-          const remotable = getValForSlot(vref);
-          assert(remotable, X`no remotable for ${vref}`);
-          // console.log(`adding ${vref} to reachableRemotables`);
-          // TODO: these should be refcounted instead of this being a journey of no return
-          reachableRemotables.add(remotable);
+          const remotable = requiredValForSlot(vref);
+          if (remotableRefCounts.has(remotable)) {
+            const oldRefCount = /** @type {number} */ (remotableRefCounts.get(
+              remotable,
+            ));
+            remotableRefCounts.set(remotable, oldRefCount + 1);
+          } else {
+            remotableRefCounts.set(remotable, 1);
+          }
         }
       } else {
-        // We track imports, to preserve their vrefs against syscall.dropImport
-        // when the Presence goes away.
-        // TODO: these should be refcounted instead of this being a journey of no return
-        reachablePresences.add(vref);
+        // We refcount imports, to preserve their vrefs against
+        // syscall.dropImport when the Presence itself goes away.
+        incRefCount(vref);
+      }
+    }
+  }
+
+  function removeReachableVref(vref) {
+    const { type, virtual, allocatedByVat } = parseVatSlot(vref);
+    if (type === 'object') {
+      if (allocatedByVat) {
+        if (virtual) {
+          decRefCount(vref);
+        } else {
+          // exported non-virtual object: Remotable
+          const remotable = requiredValForSlot(vref);
+          const oldRefCount = /** @type {number} */ (remotableRefCounts.get(
+            remotable,
+          ));
+          assert(oldRefCount > 0, `attempt to decref ${vref} below 0`);
+          if (oldRefCount === 1) {
+            remotableRefCounts.delete(remotable);
```

Comment:
I think this might not be entirely right.  `remotableRefCounts` is a regular map that holds onto a virtually referenced remotable via the key; deleting the map entry removes the key and tees up the object to be swept up normally in the next GC pass if it is now unreferenced in memory as a consequence.  The benefit from prodding `scanForDeadObjects` to do another GC pass would be that the Remotable would be swept up eagerly rather than at the end of the next crank, but this logic here is in `removeReachableVref` which is called from several places, some of which can happen during normal crank computation, not underneath `scanForDeadObjects`.  In other words, it's not necessarily in a position to prod `scanForDeadObjects` to do another loop because `scanForDeadObjects` might not be (in fact, probably isn't) running, and so there's no return path to communicate the flag.

One way to finesse this might be to have a flag inside liveSlots outside the scope of `scanForDeadObjects` that if you set it will cause `scanForDeadObjects` to loop if it sees it, but will have no effect if `scanForDeadObjects` is not running.  This flag would work like the `doMore` flag does now, but would be located in the an outer scope.  This probably means we'd have to add a function to liveSlots for setting the flag.  This seems doable, but I'd like to think about it a bit more as I'm really not eager to have subtle control flow modulated by random boolean flags set from the outside.